flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/catalog/CatalogTableITCase.scala

```diff
@@ -0,0 +1,450 @@
+    )
+    TestCollectionTableFactory.initData(sourceData)
+    val sourceDDL =
+      """
+        |create table t1(
+        |  a int,
+        |  b varchar,
+        |  c int
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val sinkDDL =
+      """
+        |create table t2(
+        |  a int,
+        |  b varchar,
+        |  c int
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |insert into t2
+        |select t1.a, t1.b, (t1.a + 1) as c from t1
+      """.stripMargin
+    tableEnv.sql(sourceDDL)
+    tableEnv.sql(sinkDDL)
+    tableEnv.sql(query)
+    execJob("testJob")
+    assertEquals(sourceData.sorted, TestCollectionTableFactory.RESULT.sorted)
+  }
+
+  @Ignore // need to implement
+  @Test
+  def testInsertTargetTableWithComputedColumn(): Unit = {
+    TestCollectionTableFactory.initData(SOURCE_DATA)
+    val sourceDDL =
+      """
+        |create table t1(
+        |  a int,
+        |  b varchar,
+        |  c int
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val sinkDDL =
+      """
+        |create table t2(
+        |  a int,
+        |  b varchar,
+        |  c as a + 1
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |insert into t2(a, b)
+        |select t1.a, t1.b from t1
+      """.stripMargin
+    tableEnv.sql(sourceDDL)
+    tableEnv.sql(sinkDDL)
+    tableEnv.sql(query)
+    execJob("testJob")
+    assertEquals(SOURCE_DATA.sorted, TestCollectionTableFactory.RESULT.sorted)
+  }
+
+  @Test
+  def testInsertWithJoinedSource(): Unit = {
+    val sourceData = List(
+      toRow(1, 1000, 2),
+      toRow(2, 1, 3),
+      toRow(3, 2000, 4),
+      toRow(1, 2, 2),
+      toRow(2, 3000, 3)
+    )
+
+    val expected = List(
+      toRow(1, 1000, 2, 1),
+      toRow(1, 2, 2, 1),
+      toRow(2, 1, 1, 2),
+      toRow(2, 3000, 1, 2)
+    )
+    TestCollectionTableFactory.initData(sourceData)
+    val sourceDDL =
+      """
+        |create table t1(
+        |  a int,
+        |  b int,
+        |  c int
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val sinkDDL =
+      """
+        |create table t2(
+        |  a int,
+        |  b int,
+        |  c int,
+        |  d int
+        |) with (
+        |  connector = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |insert into t2
+        |select a.a, a.b, b.a, b.b
+        |  from t1 a
+        |  join t1 b
+        |  on a.a = b.b
+      """.stripMargin
+    tableEnv.sql(sourceDDL)
+    tableEnv.sql(sinkDDL)
+    tableEnv.sql(query)
+    execJob("testJob")
+    assertEquals(expected.sorted, TestCollectionTableFactory.RESULT.sorted)
+  }
+
+  @Test
+  def testInsertWithAggregateSource(): Unit = {
+    if (isStreaming) {
+      return
```

Comment:
We can create a upserting collection table sink to support this case.

Reply:
I think only test append sink is enough.